monai/data/dataset.py

```diff
@@ -93,22 +97,64 @@ def __init__(self, data, transform, cache_num=sys.maxsize, cache_rate=1.0):
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
             cache_rate (float): percentage of cached data in total, default is 1.0 (cache all).
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
+            num_workers (int): the number of worker processes to use.
+                If 0 a single thread will be used. Default is 0.
         """
         if not isinstance(transform, Compose):
             transform = Compose(transform)
         super().__init__(data, transform)
         self.cache_num = min(cache_num, int(len(self) * cache_rate), len(self))
-        self._cache = list()
+        self._cache = [None] * self.cache_num
         print('Load and cache transformed data...')
-        for i in range(self.cache_num):
-            process_bar(i + 1, self.cache_num)
-            item = data[i]
-            for _transform in transform.transforms:
-                # execute all the deterministic transforms before the first random transform
-                if isinstance(_transform, Randomizable):
-                    break
-                item = apply_transform(_transform, item)
-            self._cache.append(item)
+        if num_workers > 0:
+            self._item_processed = 0
+            if mode == "thread":
+                lock = Lock()
+
+                def _update_bar(arg):
+                    self._item_processed += 1
```

Comment:
Should we add self._item_processed after lock.acquire()?
And I think you don't need to set _item_processed to self, just a local variable item_processed is enough.
Thanks.

Reply:
Yes you're right, it'll be cleaner. I'll update it

Reply:
I think the `item_processed += 1` should be in lock because it's not **atomic operation** in hardware. Check this doc: https://software.intel.com/en-us/node/506090.
Thanks.

Reply:
I have updated `self._item_processed` to be within the lock scope, thanks.
However, the variable has to be assigned to self as it would not be defined otherwise (I think it is because the function `_load_cache_item_thread` has to be pickled so I removed it from the `__init__` scope to avoid confusion).